src/wsi/sdl2/wsi_platform_sdl2.cpp

```diff
@@ -0,0 +1,69 @@
+#if defined(DXVK_WSI_SDL2)
+
+#include "wsi_platform_sdl2.h"
+#include "../../util/util_error.h"
+#include "../../util/util_string.h"
+#include "../../util/util_win32_compat.h"
+
+#include <SDL2/SDL_vulkan.h>
```

Comment:
The canonical way to include SDL2 headers is `#include "SDL.h"` (as per https://wiki.libsdl.org/SDL2/FAQDevelopment). This will require using the `-I` flags from `sdl2-config --cflags` or `pkg-config --cflags sdl2`, which in Meson means compiling the SDL2 WSI files with `dependency('sdl2')` or at least a `partial_dependency`. In this case, because of the use of `dlopen()`-style dynamic loading, you'd presumably want to use a `partial_dependency`.

This is not just a theoretical thing: in Debian, we have had some trouble in the past with interoperability with the `SDL2/SDL_whatever.h` style. It works most of the time, except for when it doesn't, and can have weird quirks dependent on whether SDL was compiled with Autotools or CMake.

(If you're planning to add a SDL3 WSI in future when SDL3 has stabilized, note that this is different in SDL3: the recommendation in https://wiki.libsdl.org/SDL3/FAQDevelopment is `<SDL3/SDL.h>`. This is an intentional interface change, and all packaged versions of SDL3 are expected to make it work.)

Reply:
That all makes sense to me - incidentally, is it possible to get the soname via meson? We'd be mimicking what SDL does for its dynamic loading support: https://github.com/libsdl-org/SDL/blob/main/cmake/sdlchecks.cmake#L546

Reply:
No, I don't know of a trivial way to get the SONAME.

Honestly, my answer would be to keep hard-coding the expected SONAME. If the ABI of SDL or GLFW changes incompatibly, then you already need to know about that and react to it, because the symbols that you list in [wsi_platform_sdl2_funcs.h](https://github.com/doitsujin/dxvk/pull/3834/files#diff-4ccb978d31117e5c2b73e6f2f3766f412b897cec609c1ad010950e38481cb0fe) and [wsi_platform_glfw_funcs.h](https://github.com/doitsujin/dxvk/pull/3834/files#diff-e109f7a3c9d6cbe0419f2dc9fce561cc2cf2cc81b814827bab5748ae73cee968) will need reviewing anyway. That's because one possible reason for an ABI break would be an incompatible change to the signature of one of those functions, for example if `GLFWmonitor **glfwGetMonitors(int *)` changed to `GLFWmonitor **glfwGetMonitors(size_t *)` in a newer version of GLFW.

The traditional build-system-developer answer, e.g. in `libtool`, would be "no you can't learn the SONAME, and it's wrong to want to", because shared libraries and `dlopen()`'able modules happen to be the same thing on popular platforms like Windows and Linux but have historically been implemented differently in some OSs. In that view of the world, each shared object is either correct to link like `-lSDL3` or correct to dlopen like `dlopen("/path/to/myplugin.so")`, but never both; and the pedantically correct approach would be for DXVK to have in-tree plugins for SDL and GLFW (no need for a stable API/ABI between library and plugin), similar to how libdecor has in-tree plugins for Cairo and GTK. In that design, the plugin would be the only thing that is dlopened, and the plugin would have a normal DT_NEEDED dependency (hard dependency) on the library it uses.

But, assuming that you have no interest in making DXVK run on AIX or HPUX or whatever other museum piece has this limitation, I can see why you would prefer to cut out the middle-man and dlopen the library itself!

It looks as though SDL's `FindLibraryAndSONAME` assumes that if you take the `basename` of the `realpath` of the shared library, and then truncate `libfoo.so.1.2.3.4` to `libfoo.so.1`, the result is the SONAME. This is not strictly correct - some libraries, confusingly, have SONAME `libfoo.so.0` and a symlink `libfoo.so.0 -> libfoo.so.1.2.3.4` - but it "often" works correctly for "normal" libraries.

I'm guessing that compiling a version of https://salsa.debian.org/gnome-team/gobject-introspection/-/blob/debian/1.78.1-14/debian/elf-get-needed.c?ref_type=tags (except with `DT_SONAME` instead of `DT_NEEDED`) for the `native` (build) architecture, and then running it against the discovered host-architecture library, is not the answer you're looking for either - but that would be the most "correct" way to do it.

Reply:
Fair enough, thanks for the info - maybe this is something we can punt off to the other SDL maintainers to solve For Real :P

Should have fixes for all of the other notes at some point this week!

Reply:
The includes have been fixed in the WSI PR as well.